src/loader/mod.rs

```diff
@@ -0,0 +1,580 @@
+    SeekProgramHeader,
+    SeekBzImageEnd,
+    SeekBzImageHeader,
+    SeekBzImageCompressedKernel,
+    SeekInitrdImage,
+}
+pub type Result<T> = std::result::Result<T, Error>;
+
+impl error::Error for Error {
+    fn description(&self) -> &str {
+        match self {
+            Error::BigEndianElfOnLittle => {
+                "Trying to load big-endian binary on little-endian machine"
+            }
+            Error::CommandLineCopy => "Failed writing command line to guest memory",
+            Error::CommandLineOverflow => "Command line overflowed guest memory",
+            Error::InvalidElfMagicNumber => "Invalid Elf magic number",
+            Error::InvalidProgramHeaderSize => "Invalid program header size",
+            Error::InvalidProgramHeaderOffset => "Invalid program header offset",
+            Error::InvalidProgramHeaderAddress => "Invalid Program Header Address",
+            Error::InvalidEntryAddress => "Invalid entry address",
+            Error::InvalidBzImage => "Invalid bzImage",
+            Error::InvalidKernelStartAddress => "Invalid kernel start address",
+            Error::InitrdImageSizeTooLarge => "Initrd image size too large",
+            Error::ReadElfHeader => "Unable to read elf header",
+            Error::ReadKernelImage => "Unable to read kernel image",
+            Error::ReadProgramHeader => "Unable to read program header",
+            Error::ReadBzImageHeader => "Unable to read bzImage header",
+            Error::ReadBzImageCompressedKernel => "Unable to read bzImage compressed kernel",
+            Error::ReadInitrdImage => "Unable to read initrd image",
+            Error::SeekKernelStart => "Unable to seek to kernel start",
+            Error::SeekElfStart => "Unable to seek to elf start",
+            Error::SeekProgramHeader => "Unable to seek to program header",
+            Error::SeekBzImageEnd => "Unable to seek bzImage end",
+            Error::SeekBzImageHeader => "Unable to seek bzImage header",
+            Error::SeekBzImageCompressedKernel => "Unable to seek bzImage compressed kernel",
+            Error::SeekInitrdImage => "Unable to seek initrd image",
+        }
+    }
+}
+
+impl Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "Kernel Loader Error: {}", Error::description(self))
+    }
+}
+
+/// * `kernel_load` - The actual `guest_mem` address where kernel image is loaded start.
+/// * `kernel_end` - The offset of `guest_mem` where kernel image load is loaded finish, return
+///                  in case of loading initrd adjacent to kernel image.
+/// * `setup_header` - The setup_header belongs to linux boot protocol, only for bzImage, vmm
+///                    will use it to setup setup_header.init_size, which is a must for bzImage
+///                    direct boot.
+#[derive(Debug, Default, Copy, Clone, PartialEq)]
+pub struct KernelLoaderResult {
+    pub kernel_load: GuestAddress,
+    pub kernel_end: GuestUsize,
```

Comment:
Why is `kernel_end` a usize while `kernel_load` is a u64?

Reply:
There will be memory alignment calculation based on the returned kernel end for the next data struct to be put in, so just return GuestUsize directly.